imap.go

```diff
@@ -1,102 +1,182 @@
+	cfg.Bool("insecure_auth", false, &insecureAuth)
+	cfg.Bool("io_debug", false, &insecureAuth)
+	cfg.Bool("debug", true, &endp.Log.Debug)
+	if _, err := cfg.Process(globalCfg, &rawCfg); err != nil {
+		return err
+	}
+
+	addresses := make([]Address, 0, len(rawCfg.Args))
+	for _, addr := range rawCfg.Args {
+		saddr, err := standardizeAddress(addr)
 		if err != nil {
-			return nil, err
+			return fmt.Errorf("imap: invalid address: %s", endp.name)
+		}
+		if saddr.Scheme != "imap" && saddr.Scheme != "imaps" {
+			return fmt.Errorf("imap: imap or imaps scheme must be used, got %s", saddr.Scheme)
 		}
+		addresses = append(addresses, saddr)
+	}
+
+	endp.serv = imapserver.New(endp)
+	endp.serv.AllowInsecureAuth = insecureAuth
+	endp.serv.TLSConfig = endp.tlsConfig
+	endp.serv.ErrorLog = endp.Log.DebugWriter()
+	if ioDebug {
+		endp.serv.Debug = os.Stderr
 	}
 
-	s := imapserver.New(be)
+	if err := endp.enableExtensions(); err != nil {
+		return err
+	}
 
-	if tokens, ok := tokens["errors"]; ok {
-		err := setIMAPErrors(caddyfile.NewDispenserTokens("", tokens), s)
+	for _, addr := range addresses {
+		var l net.Listener
+		var err error
+		l, err = net.Listen(addr.Network(), addr.Address())
 		if err != nil {
-			return nil, err
+			return fmt.Errorf("failed to bind on %v: %v", addr, err)
 		}
+		endp.Log.Printf("imap: listening on %v", addr)
+
+		if addr.IsTLS() {
+			if endp.tlsConfig == nil {
+				return errors.New("can't bind on IMAPS endpoint without TLS configuration")
+			}
+			l = tls.NewListener(l, endp.tlsConfig)
+		}
+
+		endp.listeners = append(endp.listeners, l)
+
+		endp.listenersWg.Add(1)
+		addr := addr
+		go func() {
+			if err := endp.serv.Serve(l); err != nil && !strings.HasSuffix(err.Error(), "use of closed network connection") {
+				endp.Log.Printf("imap: failed to serve %s: %s", addr, err)
+			}
+			endp.listenersWg.Done()
+		}()
+	}
+
+	if endp.serv.AllowInsecureAuth {
+		endp.Log.Println("authentication over unencrypted connections is allowed, this is insecure configuration and should be used only for testing!")
 	}
-	if _, ok := tokens["compress"]; ok {
-		s.Enable(imapcompress.NewExtension())
+	if endp.serv.TLSConfig == nil {
+		endp.Log.Println("TLS is disabled, this is insecure configuration and should be used only for testing!")
+		endp.serv.AllowInsecureAuth = true
 	}
 
-	return s, nil
+	return nil
 }
 
-func newIMAPProxy(d caddyfile.Dispenser) (imapbackend.Backend, error) {
-	for d.Next() {
-		args := d.RemainingArgs()
+func (endp *IMAPEndpoint) Name() string {
+	return "imap"
+}
 
-		if len(args) == 1 {
-			addr, err := standardizeAddress(args[0])
-			if err != nil {
-				return nil, err
-			}
+func (endp *IMAPEndpoint) InstanceName() string {
+	return endp.name
+}
 
-			target := addr.Host + ":" + addr.Port
-			if addr.IsTLS() {
-				return imapproxy.NewTLS(target, nil), nil
-			} else {
-				return imapproxy.New(target), nil
-			}
-		}
-	}
+func (endp *IMAPEndpoint) Version() string {
```

Comment:
I'm wondering whether we really need those

Reply:
We have at least one "out-of-tree" module, so version information tied to the modules can be useful for bug reports. But I guess we can really drop it and just report go-imap-sql's version in debug log.